File: src/CubeBraidSDK/AGV_SDK/agv_sdk_demo.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include "CubeBraidSDK/AGV_SDK/AGV_SDK.h"

int main() 
{
    std::cout << "[AGV Demo] 初始化 AGV 控制器..." << std::endl;
    agv_sdk::AGVController agv;

    // 连接设备
    const std::string target_ip = "192.168.1.91"; // AGV IP
    const int target_port = 5005;   // AGV 端口
    
    std::cout << "[AGV Demo] 正在连接设备: " << target_ip << ":" << target_port << "..." << std::endl;
    if (!agv.connectAGV(target_ip, target_port)) 
    {
        std::cerr << "[AGV Demo] 连接失败，请检查网络设置。" << std::endl;
        return -1;
    }
    std::cout << "[AGV Demo] 连接成功！" << std::endl;

    // 登录认证 (假设密码已为 Hash 值)
    std::cout << "[AGV Demo] 发起登录请求..." << std::endl;
    if (!agv.login("admin", "21232f297a57a5a743894a0e4a801fc3")) 
    {
        std::cerr << "[AGV Demo] 登录失败！" << std::endl;
        agv.disconnectAGV();
        return -1;
    }
    std::cout << "[AGV Demo] 登录成功！" << std::endl;

    std::this_thread::sleep_for(std::chrono::milliseconds(200));    // 200ms

    // 切换为自动控制模式
    std::cout << "[AGV Demo] 切换为自动控制模式..." << std::endl;
    agv.switchControlMode(agv_sdk::ControlMode::Auto);

    std::this_thread::sleep_for(std::chrono::milliseconds(200));    // 200ms

    // 查询当前状态与位姿
    agv.querySystemState();
    agv_sdk::AGVPose pose = agv.getPose();
    std::cout << "前进后位姿 -> 前方距离 X: " << pose.roll << "m" << std::endl;
    std::cout << "前进后位姿 -> 左侧距离 Y: " << pose.pitch << "m" << std::endl;
    // 控制移动：前进 1800mm
    std::cout << "[AGV Demo] 发送前进到底指令: 1800mm (超时 100s)..." << std::endl;
    if (agv.goForward(1800.0, 100000)) 
    {
        std::cout << "[AGV Demo] 前进到底到位！" << std::endl;
    } 
    else 
    {
        std::cerr << "[AGV Demo] 前进到底指令超时或执行失败！" << std::endl;
    }

    // 更新位姿
    pose = agv.getPose();
    std::cout << "前进后位姿 -> 前方距离 X: " << pose.roll << "m" << std::endl;
    std::cout << "前进后位姿 -> 左侧距离 Y: " << pose.pitch << "m" << std::endl;

    // 前进 500mm
    std::cout << "[AGV Demo] 发送单步前进指令: 500mm (超时 100s)..." << std::endl;
    if (agv.goBack(500.0, 100000)) 
    {
        std::cout << "[AGV Demo] 单步前进到位！" << std::endl;
    }

    // 更新位姿
    pose = agv.getPose();
    std::cout << "前进后位姿 -> 前方距离 X: " << pose.roll << "m" << std::endl;
    std::cout << "前进后位姿 -> 左侧距离 Y: " << pose.pitch << "m" << std::endl;

    // 后退 500mm
    std::cout << "[AGV Demo] 发送单步后退指令: 500mm (超时 100s)..." << std::endl;
    if (agv.goBack(-500.0, 100000)) 
    {
        std::cout << "[AGV Demo] 单步后退到位！" << std::endl;
    }

    // 更新位姿
    pose = agv.getPose();
    std::cout << "前进后位姿 -> 前方距离 X: " << pose.roll << "m" << std::endl;
    std::cout << "前进后位姿 -> 左侧距离 Y: " << pose.pitch << "m" << std::endl;

    // 切换为手动控制模式
    std::cout << "[AGV Demo] 切换为手动控制模式..." << std::endl;
    agv.switchControlMode(agv_sdk::ControlMode::Manual);

    std::cout << "[AGV Demo] 切换为手动速度控制模式: 100mm/s速度前进2000ms (自动重发与归零停止)..." << std::endl;
    // 传入 vx=100.0, vy=0.0, w=0.0，持续时间 2000ms，内部每 50ms 重发一次
    agv.moveManualForDuration(100.0f, 0.0f, 0.0f, 2000, 50);

    std::cout << "[AGV Demo] 切换为手动速度控制模式: -100mm/s速度前进2000ms (自动重发与归零停止)..." << std::endl;
    // 传入 vx=-100.0, vy=0.0, w=0.0，持续时间 2000ms，内部每 50ms 重发一次
    agv.moveManualForDuration(-100.0f, 0.0f, 0.0f, 2000, 50);

    std::cout << "[AGV Demo] 切换为手动速度控制模式: 0.1rad/s速度前进2000ms (自动重发与归零停止)..." << std::endl;
    // 传入 vx=0.0, vy=0.0, w=100.0，持续时间 2000ms，内部每 50ms 重发一次
    agv.moveManualForDuration(0.0f, 0.0f, 100.0f, 2000, 50);

    std::cout << "[AGV Demo] 切换为手动速度控制模式: -0.1rad/s速度前进2000ms (自动重发与归零停止)..." << std::endl;
    // 传入 vx=0.0, vy=0.0, w=-100.0，持续时间 2000ms，内部每 50ms 重发一次
    agv.moveManualForDuration(0.0f, 0.0f, -100.0f, 2000, 50);

    std::cout << "[AGV Demo] 手动控制结束并已停止" << std::endl;

    // 登出与断开
    std::cout << "[AGV Demo] 断开连接..." << std::endl;
    agv.logout();
    agv.disconnectAGV();

    std::cout << "[AGV Demo] 运行结束。" << std::endl;
    return 0;
}

File: scripts/AGV_SDK/agv_sdk_demo.py
import time
from agv_sdk import AGVClient, ControlMode

def main():
    # 配置信息
    IP = "192.168.1.91"
    PORT = 5005
    USER = "admin"
    PASS_HASH = "21232f297a57a5a743894a0e4a801fc3"

    # 使用 with 上下文管理器，确保退出程序时自动调用 disconnect 和 destroy 释放内存
    try:
        with AGVClient() as agv:
            print(f"正在连接 AGV [{IP}:{PORT}]...")
            if not agv.connect(IP, PORT):
                print("连接失败！")
                return
            print("连接成功。")

            print("执行登录...")
            if not agv.login(USER, PASS_HASH):
                print("登录失败！")
                return
            print("登录成功。")
            time.sleep(1)

            # 切换自动模式并走位
            print("切换至【自动模式】...")
            agv.switch_control_mode(ControlMode.AUTO)

            print("指令：前进到底1800mm...")
            if agv.go_forward(1800.0, timeout_ms=100000):
                print("前进到达指定位置")
            else:
                print("前进超时或未成功完成")

            # 读取位姿
            pose = agv.get_pose()
            print(f"当前位姿 -> X: {pose.x:.3f}m, Y: {pose.y:.3f}m, Yaw: {pose.yaw:.2f}°")

            print("指令：单步前进500mm...")
            if agv.go_back(500.0, timeout_ms=100000):
                print("前进到达指定位置")
            else:
                print("前进超时或未成功完成")

            # 读取位姿
            pose = agv.get_pose()
            print(f"当前位姿 -> X: {pose.x:.3f}m, Y: {pose.y:.3f}m, Yaw: {pose.yaw:.2f}°")

            print("指令：单步后退500mm...")
            if agv.go_back(-500.0, timeout_ms=100000):
                print("前进到达指定位置")
            else:
                print("前进超时或未成功完成")

            # 读取位姿
            pose = agv.get_pose()
            print(f"当前位姿 -> X: {pose.x:.3f}m, Y: {pose.y:.3f}m, Yaw: {pose.yaw:.2f}°")

            # 切换手动模式运动控制
            print("切换至【手动模式】...")
            agv.switch_control_mode(ControlMode.MANUAL)

            # 示例：以 vx=100 mm/s 前进，持续移动 2 秒（内部会自动以 100ms 频率持续发包，到期后自动归零停止）
            print("开始手动移动 2 秒...")
            if agv.move_manual_for_duration(vx=100.0, vy=0.0, w=0.0, duration_s=2, interval_s=0.1):
                print("手动移动完成并已安全停止")
            else:
                print("移动过程中出现异常")

            # 示例：以 vx=-100 mm/s 前进，持续移动 2 秒（内部会自动以 100ms 频率持续发包，到期后自动归零停止）
            print("开始手动移动 2 秒...")
            if agv.move_manual_for_duration(vx=-100.0, vy=0.0, w=0.0, duration_s=2, interval_s=0.1):
                print("手动移动完成并已安全停止")
            else:
                print("移动过程中出现异常")


            # 示例：以 w=0.1rad/s 前进，持续移动 2 秒（内部会自动以 100ms 频率持续发包，到期后自动归零停止）
            print("开始手动移动 2 秒...")
            if agv.move_manual_for_duration(vx=0.0, vy=0.0, w=100.0, duration_s=2, interval_s=0.1):
                print("手动移动完成并已安全停止")
            else:
                print("移动过程中出现异常")

            # 示例：以 w=-0.1rad/s 前进，持续移动 2 秒（内部会自动以 100ms 频率持续发包，到期后自动归零停止）
            print("开始手动移动 2 秒...")
            if agv.move_manual_for_duration(vx=0.0, vy=0.0, w=-100.0, duration_s=2, interval_s=0.1):
                print("手动移动完成并已安全停止")
            else:
                print("移动过程中出现异常")

            # 安全登出
            agv.logout()
            print("已安全登出。")

    except Exception as e:
        print(f"运行时发生错误: {e}")

if __name__ == "__main__":
    main()